### PR TITLE
Port remaining members of ErrObject, Interaction, SpecialDirectories

### PR DESCRIPTION
**Pull request by cston**
Add missing members of:
- Microsoft.VisualBasic.ErrObject
- Microsoft.VisualBasic.Interaction
- Microsoft.VisualBasic.FileIO.SpecialDirectories

Most of the members are ported from https://github.com/Microsoft/referencesource/tree/master/Microsoft.VisualBasic/runtime/msvbalib.

The remaining methods (below) use reflection to invoke a corresponding method that will be ported to an assembly higher in the stack:
- Interaction.{AppActivate,InputBox,MsgBox,Shell}
- SpecialDirectories.{CurrentUserApplicationData,AllUsersApplicationData}

Other changes to the ported code:
- Cleaned up comments
- Removed commented out code
- Removed attribute use: {HostProtection,SecuritySafeCritical,SecurityPermission,ResourceConsumption,ResourceExposure}
- Interaction.Command: removed EnvironmentPermission.Demand
- Interaction.Environ: removed EnvironmentPermission.{Assert,RevertAssert}
- ErrObject.Raise(): removed call to VBHost.GetWindowTitle()
- Interaction.GetObject: removed calls to Marshal.GetActiveObject()

### CODE REVIEW

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileIO/FileSystem.vb [hunk 1]
```diff
@@ -14,14 +14,18 @@ Imports System.Security
 Imports System.Text
 
 Imports Microsoft.VisualBasic.CompilerServices
+#If PLATFORM_WINDOWS Then
+Imports Microsoft.VisualBasic.CompilerServices.NativeMethods
+Imports Microsoft.VisualBasic.CompilerServices.NativeTypes
+#End If
```
Comment (cston) [inline src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileIO/FileSystem.vb]: The changes in this file are simply merging the contents of FileSystem*.vb so the differences with the reference source are clearer.

**Comment by cston** [comment]
@danmosemsft, @stephentoub, @ericstj, @jaredpar, @333fred, please review.
cc @KathleenDollard

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/Interaction.vb [hunk 2]
```diff
@@ -26,6 +126,23 @@ Namespace Microsoft.VisualBasic
...
+        Public Function InputBox(ByVal Prompt As String, Optional ByVal Title As String = "", Optional ByVal DefaultResponse As String = "", Optional ByVal XPos As Integer = -1, Optional ByVal YPos As Integer = -1) As String
+            Return DirectCast(InvokeMethod("InputBox", Prompt, Title, DefaultResponse, XPos, YPos), String)
+        End Function
+
+        Public Function MsgBox(ByVal Prompt As Object, Optional ByVal Buttons As MsgBoxStyle = MsgBoxStyle.OkOnly, Optional ByVal Title As Object = Nothing) As MsgBoxResult
+            Return DirectCast(InvokeMethod("MsgBox", Prompt, Buttons, Title), MsgBoxResult)
+        End Function
+
+        Private Function InvokeMethod(methodName As String, ParamArray args As Object()) As Object
+            Dim type As Type = type.GetType("Microsoft.VisualBasic._Interaction, Microsoft.VisualBasic, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a", throwOnError:=True)
```
Comment (cston) [inline src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/Interaction.vb]: We'll need a new helper type in Microsoft.VisualBasic. I'm assuming the type is public but with a special name, but perhaps the type should be internal.

File: src/Microsoft.VisualBasic.Core/tests/InteractionTests.cs [hunk 3]
```diff
@@ -114,6 +135,95 @@ public static IEnumerable<object[]> IIf_TestData()
...
+            // Not tested: valid arguments.
+        }
+
+        [Fact]
+        public void Environ_Index()
+        {
+            var pairs = GetEnvironmentVariables();
+            int n = Math.Min(pairs.Length, 255);
+
+            Assert.Throws<ArgumentException>(() => Interaction.Environ(0)).ToString();
```
Comment (333fred) [inline src/Microsoft.VisualBasic.Core/tests/InteractionTests.cs]: >.ToString() [](start = 74, length = 11)

What's up with these tostring calls?

Comment (333fred) [reply src/Microsoft.VisualBasic.Core/tests/InteractionTests.cs]: And below.

---
In reply to: [314891998](https://github.com/dotnet/corefx/pull/40365#discussion_r314891998) [](ancestors = 314891998)

**Review by 333fred — APPROVED** [review]
LGTM with the test ToString clarification.

File: src/Microsoft.VisualBasic.Core/tests/InteractionTests.cs (hunk 3 above)
Comment (cston) [reply src/Microsoft.VisualBasic.Core/tests/InteractionTests.cs]: Exception.ToString() is called to verify the exception message is constructed successfully. Added a comment and '_ =' as we discussed.

---
In reply to: [314892290](https://github.com/dotnet/corefx/pull/40365#discussion_r314892290) [](ancestors = 314892290,314891998)

**Comment by cston** [comment]
@danmosemsft, @stephentoub, @ericstj, @jaredpar, please review, thanks.

**Comment by stephentoub** [comment]
Are there any changes to the code that you ported?

**Comment by cston** [comment]
Added the set of changes from the reference source to the PR description.

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/Interaction.vb (hunk 2 above)
Comment (stephentoub) [reply src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/Interaction.vb]: I assume this new type is being added concurrently with this change and will be in 3.0?

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileIO/FileSystem.vb [hunk 4]
```diff
@@ -1137,7 +1141,20 @@ Namespace Microsoft.VisualBasic.FileIO
...
+                                    sourceFileFullPath, destinationFileFullPath, m_MOVEFILEEX_FLAGS)
+                            ' GetLastWin32Error has to be close to PInvoke call. FxCop rule.
+                            If Not succeed Then
+                                ThrowWinIOError(System.Runtime.InteropServices.Marshal.GetLastWin32Error())
+                            End If
+                        Catch
+                            Throw
+                        End Try
+#Else
+                        Throw New PlatformNotSupportedException()
```
Comment (stephentoub) [inline src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/FileIO/FileSystem.vb]: When will this ever be hit?

**Review by stephentoub — APPROVED** [review]

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/Interaction.vb (hunk 2 above)
Comment (cston) [reply src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/Interaction.vb]: The `_Interaction` type will be added to Microsoft.VisualBasic.dll in dotnet/winforms. Both changes are post-3.0. I'll hold off merging this PR until that type is available.

Comment (ericstj) [reply src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/Interaction.vb]: Are you OK with a `TypeLoadException` when someone calls this from a console app that doesn't have that type?  You might want to remap that to something else (like `PlatformNotSupportedException`).

Comment (KathleenDollard) [reply src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/Interaction.vb]: Seems that avoiding the `TypeLoadException` is desirable.

File: src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/Interaction.vb [hunk 5]
```diff
@@ -26,6 +129,23 @@ Namespace Microsoft.VisualBasic
...
+
+        Public Function MsgBox(ByVal Prompt As Object, Optional ByVal Buttons As MsgBoxStyle = MsgBoxStyle.OkOnly, Optional ByVal Title As Object = Nothing) As MsgBoxResult
+            Return DirectCast(InvokeMethod("MsgBox", Prompt, Buttons, Title), MsgBoxResult)
+        End Function
+
+        Private Function InvokeMethod(methodName As String, ParamArray args As Object()) As Object
+            Dim type As Type = type.GetType("Microsoft.VisualBasic._Interaction, Microsoft.VisualBasic, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a", throwOnError:=False)
+            Dim method As MethodInfo = type?.GetMethod(methodName)
+            If method Is Nothing Then
+                Throw New PlatformNotSupportedException()
```
Comment (ericstj) [inline src/Microsoft.VisualBasic.Core/src/Microsoft/VisualBasic/Interaction.vb]: You may want an error message here that indicates what might be done to fix the problem.

**Review by ericstj — APPROVED** [review]

**Comment by filipnavara** [comment]
These tests are now bundled in the the CoreFX test suite that is run by Mono. The check for presence of registry does not work on macOS/Linux, obviously:

```csharp
        private static readonly bool s_SupportRegistry = !PlatformDetection.IsUap;
```

/cc @ViktorHofer @marek-safar (related to https://github.com/mono/mono/pull/16681)

**Comment by danmosemsft** [comment]
@filipnavara is there something special about VB that is particularly relevant to Mono?

**Comment by filipnavara** [comment]
@danmosemsft Not necessarily. We just run all the published CoreFX tests and these were recently added to published feed.

**Comment by marek-safar** [comment]
@danmosemsft I think what Filip meant is that the check for `s_SupportRegistry` is wrong as it does not match CoreFX Registry implementation where it throws PNSE based on different conditions (https://github.com/dotnet/corefx/blob/master/src/Microsoft.Win32.Registry/src/Microsoft.Win32.Registry.csproj#L96)

**Comment by danmosemsft** [comment]
ah, was just curious.

**Comment by ericstj** [comment]
Curious how these pass our CI for linux?  @filipnavara can you open an issue that lists the specific tests which are failing?

**Comment by filipnavara** [comment]
@ericstj I filed an issue. I'm also curious why it didn't trigger an error on CI. It is possible that the list of tests run on CI differs from what Mono consumes through the feed with prebuilt assemblies.

**Pull request merged**